### PR TITLE
WOETH: Withdraw OETH in surplus.

### PR DESCRIPTION
**Pull request by clement-ux**
Allow the governor to withdraw the OETH that has been sent in surplus to the wrapper.

## Code Change Checklist

To be completed before internal review begins:

- [ ]  The contract code is complete
- [ ]  Executable deployment file
- [ ]  Fork tests that test after the deployment file runs
- [ ]  Unit tests *if needed
- [ ]  The owner has done a [full checklist review](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) of the code + tests

Internal review:

- [ ] Two approvals by internal reviewers

## Deploy checklist

Two reviewers complete the following checklist:

```
- [ ] All deployed contracts are listed in the deploy PR's description
- [ ] Deployed contract's verified code (and all dependencies) match the code in master
- [ ] The transactions that interacted with the newly deployed contract match the deploy script.
- [ ] Governance proposal matches the deploy script
- [ ] Smoke tests pass after fork test execution of the governance proposal
```

### CODE REVIEW

File: contracts/contracts/token/WOETH.sol [hunk 1]
```diff
@@ -80,9 +80,11 @@ contract WOETH is ERC4626, Governable, Initializable {
         external
         onlyGovernor
     {
-        //@dev TODO: we could implement a feature where if anyone sends OETH direclty to
-        // the contract, that we can let the governor transfer the excess of the token.
-        require(asset_ != address(asset()), "Cannot collect OETH");
+        if (asset_ == address(asset())) {
+            uint256 surplus = OETH(asset()).balanceOf(address(this)) - totalAssets();
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:84]: On a quick look, I'm a bit lost here. OETH is a rebasing token. wOETH's appreciation comes from the OETH yields (which are distributed during rebases with increasing balances). Won't this forgo all unrealized yields as well?

Comment (shahthepro) [reply contracts/contracts/token/WOETH.sol:84]: Looks like it has an internal log of oethCredits (updating during wrap and unwrap) but let me check

File: contracts/test/token/woeth.js [hunk 2]
```diff
@@ -105,7 +106,18 @@ describe("WOETH", function () {
       await expect(woeth).to.have.approxBalanceOf("150", oeth);
       await expect(woeth).to.have.a.totalSupply("50");
     });
-
+    it("should decrease with an OETH negative rebase", async () => {
```
Comment (sparrowDom) [inline contracts/test/token/woeth.js:109]: We [only rebase upwards](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/vault/VaultCore.sol#L389-L391). If the protocol incurs a loss the rebases are "paused" until enough capital is earned for a next positive rebase

File: contracts/contracts/token/WOETH.sol [hunk 3]
```diff
@@ -80,9 +80,11 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+            require(amount_ <= surplus, "Cannot collect OETH more than surplus");
```
Comment (sparrowDom) [inline contracts/contracts/token/WOETH.sol:85]: This is really nitpicky but Solidity is much more effective in handling strings when [they are 31 characters long or less ](https://ethereum.stackexchange.com/questions/107282/storage-and-memory-layout-of-strings). It uses 1 slot instead of 3. 

If you don't mind can you rephrase the error message to a 31 or less char long one?

File: contracts/contracts/token/WOETH.sol [hunk 4]
```diff
@@ -95,7 +97,8 @@ contract WOETH is ERC4626, Governable, Initializable {
      * @return amount of OETH credits the OETH amount corresponds to
      */
     function _creditsPerAsset(uint256 oethAmount)
-        internal
+        internal 
+        view
```
Comment (sparrowDom) [inline contracts/contracts/token/WOETH.sol:101]: nice catch 👍

File: contracts/contracts/token/WOETH.sol (hunk 1 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:84]: With [WOETH ignore donations PR](https://github.com/OriginProtocol/origin-dollar/pull/2106/files#diff-8bc11d78a06b7abf6bfae9f732a591fbd71cac02a935b1bc3f57ee04a22d8f41R119) WOETH keeps the balance of the internal credits (of the OETH token) that change on mint, deposit, redeem, withdraw. 

I think this approach is mathematically correct since the `surplus` figures out the difference between WOETH internal credits amount and OETH internal credits (for WOETH contract) multiplied by the credits per token - in other words operating with OETH token amounts. 

Surplus could also be calculated using another approach by finding the difference of internal credits tracked by WOETH and OETH and then multiplying it by the current credits per token ( to arrive to the OETH token surplus amount). 

I think in either of the approaches the un-realized rebase yields aren't left on the table, since as long as we do the correct math on the underlying credits per token. The rebase will take care of correct yield distribution.

Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:84]: Still a spot on comment @shahthepro 🙏  These are the sort of checks we need to make to be sure the math works correctly.

We could add a test for this @clement-ux, where: 
- WOETH has initial supply of 100 OETH
- 2 OETH are transferred to the WOETH contract
- OETH rebases (say increases balance of OETH on WOETH contract by x2)
- governor should be able to transfer out 4 OETH and no more

**Review by sparrowDom — COMMENTED** [review]
Left a couple of comments, some minor changes and this PR is good to go. 

Great job with it @clement-ux  🙏

File: contracts/contracts/token/WOETH.sol (hunk 3 above)
Comment (clement-ux) [reply contracts/contracts/token/WOETH.sol:85]: Oh yes sure, nice catch!

File: contracts/contracts/token/WOETH.sol (hunk 1 above)
Comment (shahthepro) [reply contracts/contracts/token/WOETH.sol:84]: Thanks for the detailed explanation. Will check both PRs today to get more clarity on this.

Yeah, totally agree that we should be having unit and fork tests with different scenarios involving donations and rebases.

Comment (clement-ux) [reply contracts/contracts/token/WOETH.sol]: I've added the test that @sparrowDom suggested in this [commit](https://github.com/OriginProtocol/origin-dollar/pull/2119/commits/e192d5175c592645ddf40e0a070f0b0d072aa877) and it passes.

Comment (clement-ux) [reply contracts/contracts/token/WOETH.sol]: > Surplus could also be calculated using another approach by finding the difference of internal credits tracked by WOETH and OETH and then multiplying it by the current credits per token ( to arrive to the OETH token surplus amount).

Do you mean dividing instead of multiplying maybe? 

An implementation could look something like this: 
``` solidity
(uint256 creditsBalanceHighres, uint256 creditsPerTokenHighres, ) = OETH(asset()).creditsBalanceOfHighres(address(this));
uint256 surplus = (creditsBalanceHighres - oethCreditsHighres).divPrecisely(creditsPerTokenHighres);
```

This could consume less gas but can lead to rounding issues IMO.

File: contracts/test/token/woeth.js (hunk 2 above)
Comment (clement-ux) [reply contracts/test/token/woeth.js:109]: I was thinking of adding it so that the wrapper would continue to work even if the OETH implementation changed and allowed negative rebasing. But it's true that if the implementation changes, there will probably be stuff to modify in the wrapper anyway, so no need to create tests for a scenario that isn't at the moment ahah.

I'm removing it!

**Review by sparrowDom — APPROVED** [review]
LGTM

**Comment by clement-ux** [comment]
Not relevant anymore due to new WOETH implementation.
